File: archive/2020f/solutions/TanyasRevenge.java
import java.util.*;
import java.io.*;

// By Ayush Varshney
public class TanyasRevenge {
	static int[][] up, down;
	static ArrayList<Integer>[] adj;
	static boolean[] red;
	static int n;

	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		red = new boolean[n];
		String s = sc.next();
		for (int i = 0; i < n; i++) {
			red[i] = s.charAt(i) == '1';
		}
		adj = new ArrayList[n];
		for (int i = 0; i < n; i++) {
			adj[i] = new ArrayList<>();
		}
		for (int i = 0; i < n - 1; i++) {
			int a = sc.nextInt() - 1, b = sc.nextInt() - 1;
			adj[a].add(b);
			adj[b].add(a);
		}
		up = new int[n][n + 1];
		down = new int[n][n + 1];
		System.out.println(dfs(0, -1));
		sc.close();
	}

	static int dfs(int u, int par) {
		for (int child : adj[u]) {
			if (child != par) {
				dfs(child, u);
			}
		}
		if (red[u]) {
			for (int k = 0; k < n; k++) {
				up[u][k] = k;
				down[u][k] = k;
				for (int child : adj[u]) {
					if (child != par) {
						up[u][k] += Math.max(up[child][k + 1], down[child][1]);
						down[u][k] += Math.max(up[child][1], down[child][k + 1]);
					}
				}
			}
		} else {
			for (int k = 0; k < n; k++) {
				up[u][k] = k;
				down[u][k] = 0;
				for (int child : adj[u]) {
					if (child != par) {
						up[u][k] += Math.max(up[child][k], down[child][1]);
						down[u][k] += Math.max(up[child][0], down[child][k + 1]);
					}
				}
			}
		}
		return up[0][0];
	}

}

File: archive/2020f/solutions/TanyasRevenge.py
from sys import stdin, stdout, setrecursionlimit

setrecursionlimit(10000)

n = int(stdin.readline())
red = [x=="1" for x in stdin.readline()]

adj = [[] for i in range(n)]
for i in range(n-1):
    inp = stdin.readline().split()
    a, b = int(inp[0])-1, int(inp[1])-1
    adj[a].append(b)
    adj[b].append(a)

up = [[0]*(n+1) for i in range(n)]
down = [[0]*(n+1) for i in range(n)]

def dfs(u, par):
    for child in adj[u]:
        if child != par:
            dfs(child, u)
            
    if red[u]:
        for k in range(n):
            up[u][k]= k
            down[u][k] = k
            for child in adj[u]:
                if child != par:
                    up[u][k] += max(up[child][k+1], down[child][1])
                    down[u][k] += max(up[child][1], down[child][k+1])
    else:
        for k in range(n):
            up[u][k]= k
            down[u][k] = 0
            for child in adj[u]:
                if child != par:
                    up[u][k] += max(up[child][k], down[child][1])
                    down[u][k] += max(up[child][0], down[child][k+1])

dfs(0, -1)
stdout.write(str(up[0][0]))
